### PR TITLE
Add packet-io support

### LINKED ISSUE #26: Trying to handle packets between data plane and control plane using p4runtime-shell
I am trying to send and receive packets between the switch and the CPU, i.e communication between data plane and control plane.

On the switch I am running **bmv2 with stratum**. I am working with the [opennetworking/mn-stratum](https://hub.docker.com/r/opennetworking/mn-stratum) docker image.

On the other side, I am using the [p4lang/p4runtime-shell](https://hub.docker.com/r/p4lang/p4runtime-sh) docker image to load pipeline configuration and push table entries.

As far as I know, in order to send packets to the CPU, I just need to forward packets to the egress port that corresponds to the CPU port specified in the `--cpu port` argument when running the bmv2 switch. So that's what I do in my p4 program in the ingress processing:

```p4
standard_metadata.egress_spec = CPU_PORT;
```

But now I want to handle this packet in the CPU and I do not know really how to do it. I don't even know what happens when I send the packet, because I'm not able to see the bmv2 log properly (already asked here https://github.com/stratum/stratum/issues/21).

Also, I do not know how to send packets from the control plane to the data plane using p4runtime-shell.

So I am a bit lost here. Any clues about how can I handle in/out packets?

### PR DESCRIPTION
**Pull request by Yi-Tseng**
This pull request adds packet-io support.
Also, modify the way we handle stream response messages to support other types of messages.

`usage/packet_io.md` include the way to use the packet-io

fixes #26

### CODE REVIEW

File: p4runtime_sh/p4runtime.py [hunk 1]
```diff
@@ -163,13 +169,23 @@ def stream_recv_wrapper(stream):
             @parse_p4runtime_error
             def stream_recv():
                 for p in stream:
-                    self.stream_in_q.put(p)
+                    if p.HasField("arbitration"):
+                        self.stream_in_q["arbitration"].put(p)
+                    elif p.HasField("packet"):
+                        self.stream_in_q["packet"].put(p)
+                    elif p.HasField("digest"):
+                        self.stream_in_q["digest"].put(p)
```
Comment (Yi-Tseng) [inline p4runtime_sh/p4runtime.py:177]: I'll add support for digest in the next PR, so I just leave this here

**Review by antoninbas — COMMENTED** [review]
Thanks for working on this. There's no unit test?

I think this is a good first step, and I have no issue merging this. In the future several improvements are possible IMO: ability to disable packet-ins to avoid growing queue size, ability to filter packet-ins, ability to apply a lambda to packet-ins, etc.

File: p4runtime_sh/shell.py [hunk 2]
```diff
@@ -2263,6 +2264,106 @@ def add(self, egress_port=None, instance=0):
...
+
+    def set(self, **kwargs):
+        for name, value in kwargs.items():
+            self[name] = value
+
+    def clear(self):
+        self._md.clear()
+
+
+class PacketIO(_P4EntityBase):
```
Comment (pudelkoM) [inline p4runtime_sh/shell.py:2322]: Could this class be split up into `in` and `out`?

Seems weird to me to be able to call `send` on a `PacketIn`.

File: p4runtime_sh/p4runtime.py [hunk 3]
```diff
@@ -198,17 +214,18 @@ def handshake(self):
...
         start = time.time()
         try:
             while True:
                 remaining = timeout - (time.time() - start)
                 if remaining < 0:
                     break
-                msg = self.stream_in_q.get(timeout=remaining)
+                msg = self.stream_in_q[type_].get(timeout=remaining)
                 if msg is None:
                     return None
```
Comment (pudelkoM) [inline p4runtime_sh/p4runtime.py:228]: The loop can be removed as we sort stream messages by type beforehand.

**Comment by pudelkoM** [comment]
Thanks for the review @antoninbas!

> ability to apply a lambda to packet-ins

Do you envision a global callback (then invoked in the background) or something more ad-hoc like with `table_entry.read(lambda e: print(e))`?

>ability to disable packet-ins to avoid growing queue size

I guess a global per-queue lambda would also solve the growing queue size problem. The default would just be a discarding function.

>ability to filter packet-ins

Filtering on metadata fields?

**Comment by antoninbas** [comment]
@pudelkoM maybe a super generic mechanism may be overkill, I think typical use cases would be:
* retrieve a single packet (currently supported by this PR)
* discard all packets
* write all packets to a file in the background
* print all packets in a loop until I hit Ctrl-C: this can be achieved with the current implementation from @Yi-Tseng so nothing is needed here besides an example in `usage/` maybe

Originally I thought we would be able to register custom handlers which would apply to all packet-ins or packet-ins matching a specific filter (based on metadata fields), but that may be out-of-scope for a shell. Although the ability to redirect to different files (3rd use case above) based on the value of certain metadata fields could be neat.

**Comment by Yi-Tseng** [comment]
Hi @antoninbas, thanks for the review
I totally agree that we can add support to register some call-back functions/handlers to handle packet-ins. I should be able to include it in this PR.

File: README.md [hunk 4]
```diff
@@ -93,6 +93,8 @@ about P4Info objects.
 P4Runtime `Entity` fields), along with `multicast_group_entry` and
 `clone_session_entry`, for runtime entity programming.
 
+`packet` is the command for packet IO, see the [usage](usage/packet_io.md) for more information.
```
Comment (antoninbas) [inline README.md:96]: needs to be updated now that we have different commands for `packet_in` / `packet_out`

File: p4runtime_sh/context.py [hunk 5]
```diff
@@ -55,6 +56,11 @@ class P4RuntimeEntity(enum.Enum):
     packet_replication_engine_entry = 8
 
 
+@enum.unique
+class P4RuntimeStreamMessage(enum.Enum):
+    packet = 1
```
Comment (antoninbas) [inline p4runtime_sh/context.py:61]: you don't seem to be using this anymore?

File: p4runtime_sh/shell.py [hunk 6]
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
...
+            self._md_info[md.name] = md
+            self._md[md.name] = self._parse_md('0', md)
+        self._set_docstring()
+
+    def _set_docstring(self):
+        self.__doc__ = "Available metadata:\n\n"
+        for name, info in self._md_info.items():
+            self.__doc__ += str(info)
+        self.__doc__ += """
+Set a metadata value with <self>['<metadata_name>'] = '...'
```
Comment (antoninbas) [inline p4runtime_sh/shell.py:2298]:
```suggestion
Set a metadata value with <self>.['<metadata_name>'] = '...'
```

File: p4runtime_sh/shell.py [hunk 7]
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
...
+        self._md.clear()
+
+    def values(self):
+        return self._md.values()
+
+
+latest_hid = 0
+packet_in_handles = {}  # Handle ID -> handle function
+class PacketIn():
+
```
Comment (antoninbas) [inline p4runtime_sh/shell.py:2344]: extra empty whiteline?

File: p4runtime_sh/shell.py [hunk 8]
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
...
+    def clear(self):
+        self._md.clear()
+
+    def values(self):
+        return self._md.values()
+
+
+latest_hid = 0
+packet_in_handles = {}  # Handle ID -> handle function
```
Comment (antoninbas) [inline p4runtime_sh/shell.py:2342]: could we wrap this in a class to and have a single global object to manage handlers?

File: usage/packet_io.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+x = packet_in.sniff(to_file='/tmp/pkt-in.txt')
+
+# To show packet-in messages, set `verbose` to `True`
+x = packet_in.sniff(verbose=True)
+```
+
+### Set up handlers to handle different types of packet-in
+
+To handle the packet-in asynchronously, we can create handlers/functions to process
+packet-ins once we received them.
```
Comment (antoninbas) [inline usage/packet_io.md:107]: s/once we received them./as we receive them.

File: usage/packet_io.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+```python
+# Create a handle to print every packet-in messages
+P4Runtime sh >>> def handle(x):
+            ...:     print(x)
+P4Runtime sh >>> packet_in.register_handler(handle)
```
Comment (antoninbas) [inline usage/packet_io.md:113]: s/handle/handler ?

File: p4runtime_sh/shell.py [hunk 11]
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
...
+                    if md.value != filters[md.metadata_id]:
+                        return
+            if not re.search(payload_regex, packet_in_msg.payload):
+                return
+            handle(packet_in_msg)
+
+        latest_hid += 1
+        hid = latest_hid
+        packet_in_handles[hid] = _handle_wraper
+        return hid
```
Comment (antoninbas) [inline p4runtime_sh/shell.py:2398]: nit: in addition to returning the handler id, maybe we could add an attribute to the handler function object (`handler.hid = <hid>`) to make unregistering the handler easier

File: p4runtime_sh/shell.py [hunk 12]
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
...
+                        handle(msg.packet)
+                    except Exception as e:
+                        print(e)
+
+        self.recv_t = Thread(target=_packet_in_recv_func, args=(self, ))
+        self.recv_t.start()
+
+    def register_handler(self, handle, payload_regex=b'', **kwargs):
+        # Register a packet-in handle, will return the handle ID.
+        # Use keyword argument as additional filter to filter the packet in
```
Comment (antoninbas) [inline p4runtime_sh/shell.py:2370]: why "handle" instead of "handler"?

File: usage/packet_io.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+There are two ways to handle packet-in messages:
+
+### Use `sniff` function to get packet-ins
+
+The `sniff` function will return a list which contains packet-in messages when:
+
+- Receive enough number packets(by passing the `count` parameter)
+- Expired(by passing the `timeout` parameter)
+- Keyboard interrupt(Ctrl + C)
```
Comment (antoninbas) [inline usage/packet_io.md:81]:
```suggestion
- Enough packets have been received (based on the `count` parameter)
- The timeout expires (based on the `timeout` parameter)
- A keyboard interrupt occurs (Ctrl + C)
```

File: usage/packet_io.md [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+```python
+P4Runtime sh >>> packet_in.sniff(timeout=1, count=1)
```
Comment (antoninbas) [inline usage/packet_io.md:84]: maybe this should return an iterator?

File: usage/packet_io.md [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+# Will return a list once it reach the timeout time or receive enough packets
+Out[2]:
+[payload: "AAAA"
+ metadata {
+   metadata_id: 1
+   value: "\000\001"
+ }]
+
+# Both timeout and count can be "None".
+# It will wait until user send a keyboard interrupt(Ctrl + C) to the shell.
```
Comment (antoninbas) [inline usage/packet_io.md:94]: s/send/sends

File: usage/packet_io.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+P4Runtime sh >>> packet_in.sniff(timeout=None, count=None)
+
+# To log packet-in to a file, use `to_file` parameter
+x = packet_in.sniff(to_file='/tmp/pkt-in.txt')
+
+# To show packet-in messages, set `verbose` to `True`
+x = packet_in.sniff(verbose=True)
```
Comment (antoninbas) [inline usage/packet_io.md:101]: not sure `verbose` is necessary. If we use an iterator instead, user can write:
```
for x in packet_in.sniff():
    print(x)
```
making it similar to how support read operations for entities

File: usage/packet_io.md [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+```
+
+Also, we can add filters to ignore packet-in messages which we don't want to handle.
+
+```python
+# Register the handle to handle packet-in messages that contains any string
+# matches "AA.*A" in payload with ingress port 0, for example "AAbA".
+P4Runtime sh >>> def handle(x):
+            ...:     print(x)
+P4Runtime sh >>> packet_in.register_handler(handle, payload_regex=b'AA.*A', ingress_port='0')
```
Comment (antoninbas) [inline usage/packet_io.md:132]: I believe the idea was to support filtering based on controller metadata fields primarily

**Comment by Yi-Tseng** [comment]
Sorry for the delay, 
I resolved some comments and simplify the PacketIn class, the sniff function is more like the read function from other class (returns an iterator by default, and can also pass a function)

File: p4runtime_sh/shell.py [hunk 18]
```diff
@@ -616,6 +633,7 @@ def __init__(self, entity_type, p4runtime_cls, modify_only=False):
         self._entity_type = entity_type
         self._entry = p4runtime_cls()
         self._modify_only = modify_only
+        self._p4runtime_cls = p4runtime_cls
```
Comment (antoninbas) [inline p4runtime_sh/shell.py:636]: Is `self._p4runtime_cls` used anywhere?

File: p4runtime_sh/test.py [hunk 19]
```diff
@@ -1109,6 +1110,48 @@ def test_global_options_invalid(self):
...
+        md.metadata_id = 1
+        md.value = b'\x00\x01'
+        expected_msg.packet.metadata.append(md)
+
+        packet_out = sh.PacketOut()
+        packet_out.payload = b'Random packet-out payload'
+        packet_out.metadata['egress_port'] = '1'
+        packet_out.send()
+
+        actual_msg = sh.client.stream_out_q.get()
```
Comment (antoninbas) [inline p4runtime_sh/test.py:1152]: how can this work consistently? I would expect the thread executing `stream_req_iterator` in `P4RuntimeClient` to potentially remove the item from the queue (and send it to the server) before we can retrieve it here. Isn't there a potential race condition?

Comment (Yi-Tseng) [reply p4runtime_sh/test.py:1152]: Right, I think we should store the packet-out message in the mock server(Servicer) and get the message from there.

File: p4runtime_sh/test.py [hunk 20]
```diff
@@ -67,6 +70,9 @@ def StreamChannel(self, request_iterator, context):
                 rep.arbitration.CopyFrom(req.arbitration)
                 rep.arbitration.status.code = code_pb2.OK
                 yield rep
+            elif req.HasField('packet'):
+                self.stored_packet_out.put(req)
+                yield p4runtime_pb2.StreamMessageResponse()
```
Comment (antoninbas) [inline p4runtime_sh/test.py:75]: I don't think this yield statement is needed / makes sense. When a P4Runtime server receives a PacketOut message, it doesn't need to reply with an empty stream message response.

**Review by antoninbas — COMMENTED** [review]
one more nit, otherwise LGTM

File: p4runtime_sh/test.py (hunk 20 above)
Comment (Yi-Tseng) [reply p4runtime_sh/test.py:75]: Right, I think I put this here by mistake

**Review by antoninbas — APPROVED** [review]

**Pull request merged**